cli/azd/pkg/infra/provisioning/bicep/bicep_provider.go

```diff
@@ -645,31 +645,42 @@ func latestCompletedDeployment(
 	return nil, fmt.Errorf("no deployments found for environment %s", envName)
 }
 
-// resourceGroupsFromDeployment returns the names of all the unique set of resource group name names resource groups from
-//
-//	the OutputResources section of a ARM deployment.
-func resourceGroupsFromDeployment(deployment *armresources.DeploymentExtended) []string {
-
+// resourceGroupsToDelete collects the resource groups from an existing deployment which should be removed as part of a
+// destroy operation.
+func resourceGroupsToDelete(deployment *armresources.DeploymentExtended) []string {
 	// NOTE: it's possible for a deployment to list a resource group more than once. We're only interested in the
 	// unique set.
 	resourceGroups := map[string]struct{}{}
 
-	for _, resourceId := range deployment.Properties.OutputResources {
-		if resourceId != nil && resourceId.ID != nil {
-			resId, err := arm.ParseResourceID(*resourceId.ID)
-			if err == nil && resId.ResourceGroupName != "" {
-				resourceGroups[resId.ResourceGroupName] = struct{}{}
+	if *deployment.Properties.ProvisioningState == armresources.ProvisioningStateSucceeded {
+		// For a successful deployment, we can use the output resources property to see the resource groups that were
+		// provisioned from this.
+		for _, resourceId := range deployment.Properties.OutputResources {
+			if resourceId != nil && resourceId.ID != nil {
+				resId, err := arm.ParseResourceID(*resourceId.ID)
+				if err == nil && resId.ResourceGroupName != "" {
+					resourceGroups[resId.ResourceGroupName] = struct{}{}
+				}
 			}
 		}
-	}
-
-	var resourceGroupNames []string
+	} else {
+		// For a failed deployment, the `outputResources` field is not populated. Instead, we assume that any resource
```

Comment:
I'd be fine if we just want to always union the two. I think the logic here to extract rg-level deployments makes a ton of sense without a truthful representation from ARM.

Also, does all this handle failed rg-level deployments as well?

Reply:
Going to stick with what I have here for now and we can revise based on our discussions with the control plane team (however, their suggestion to consider using deployment stacks here may remove the need for all of this code since they are now building the experience we were trying to give in a first class way).